Add initial documentation for Flink connector

matt-kwong opened the pull request:
This PR documents how `PubSubSink` works in `flink-connector/docs/content/docs/connectors/datastream/pubsub.md`, loosely following the documentation structure used by similar Flink connectors. Feel free to suggest changes to the structure.

clmccart commented on README.md:
```diff
@@ -12,6 +12,10 @@ available are:
   A reference implementation for how to use ordering keys effectively.
+* UNDER DEVELOPMENT -
+  [Flink Connector](https://github.com/GoogleCloudPlatform/pubsub/tree/master/flink-connector):
+  Send and receive messages from [Apache Flink](https://flink.apache.org/).
```
nit: "to/from"

matt-kwong replied:
Done.

clmccart commented on flink-connector/README.md:
```diff
@@ -0,0 +1,51 @@
+
+The resulting jars can be found in the `target` directory of the respective
+module.
+
```
let's add a note here on what the dependency users need to add to their pom might look like

matt-kwong replied:
Done.

clmccart commented on flink-connector/docs/content/docs/connectors/datastream/pubsub.md:
````diff
@@ -0,0 +1,158 @@
+```java
+PubSubSource.<OutputT>builder()
+    // Allow up to 10,000 message deliveries per checkpoint interval.
+    .setMaxOutstandingMessagesCount(10_000L)
````
do newer java versions still support underscores in numeric literals? (genuine question; i don't know the answer)

matt-kwong replied:
I can't find any evidence of support being removed. This should be okay as long as the Java version is 7 or later.

clmccart commented on flink-connector/docs/content/docs/connectors/datastream/pubsub.md:
```diff
@@ -0,0 +1,158 @@
+  [Pub/Sub quotas and limits](https://cloud.google.com/pubsub/quotas) for all
+  restrictions.
+
+<!-- TODO(matt-kwong) Add threading details. -->
```
is this gonna be in a follow-up PR?

matt-kwong replied:
Yes, I'm planning to expose additional builder options that influence threading.

clmccart commented on flink-connector/docs/content/docs/connectors/datastream/pubsub.md:
```diff
@@ -0,0 +1,158 @@
+
+<!-- TODO(matt-kwong) Add auth details. -->
+
+<!-- TODO(matt-kwong) Add integration testing details. -->
```
once submitted, the info in the description of #355 would be helpful to have here

matt-kwong replied:
Done.

clmccart commented on flink-connector/docs/content/docs/connectors/datastream/pubsub.md:
```diff
@@ -0,0 +1,158 @@
+    <tr>
+        <td>setCredentials(Credentials credentials)</td>
+        <td>(none)</td>
+        <td>The credentials attached to requests sent to Google Cloud Pub/Sub. The identity in the credentials must be authorized to pull messages from the subscription. If not set, then Pub/Sub source attempts to use environment-configured credentials.</td>
```
GOOGLE_APPLICATION_CREDENTIALS env var? if so, let's specify that

matt-kwong replied:
There are a few ways other ways to authenticate besides the GOOGLE_APPLICATION_CREDENTIALS env var. I added a link to https://cloud.google.com/docs/authentication/application-default-credentials to explain how default credentials work.

clmccart commented on flink-connector/docs/content/docs/connectors/datastream/pubsub.md:
```diff
@@ -0,0 +1,158 @@
+
+<!-- TODO(matt-kwong) Add sink details. -->
+
+<!-- TODO(matt-kwong) Add auth details. -->
```
would be good to have something here. even if it just point to https://github.com/apache/flink-connector-gcp-pubsub/blob/main/docs/content/docs/connectors/datastream/pubsub.md#google-credentials

matt-kwong replied:
Added a small section under Pub/Sub source. I'm not yet sure whether to add an auth section to both source and sink or make it separate from both.

clmccart replied:
since auth shouldnt behave differently in the source vs the sink, let's make it seperate

matt-kwong replied:
Done.

clmccart reviewed (changes requested):
maybe out of scope for this PR but do we have an example pipeline we can add here?

matt-kwong commented:
> maybe out of scope for this PR but do we have an example pipeline we can add here?

I added an example pipeline under `flink-connector/flink-examples-gcp-pubsub/`. I haven't referenced it in the docs, following the precedence of other connectors' docs.

clmccart commented:
> > maybe out of scope for this PR but do we have an example pipeline we can add here?
> 
> I added an example pipeline under `flink-connector/flink-examples-gcp-pubsub/`. I haven't referenced it in the docs, following the precedence of other connectors' docs.

let's reference it. I think it'd be helpful to have here (no sense hiding it)

matt-kwong commented:
> let's reference it. I think it'd be helpful to have here (no sense hiding it)

Added.

I also added documentation for the sink and integration testing. The remaining TODO will be addressed in a future PR.
